src/Microsoft.AspNetCore.Razor/CodeGenerators/Visitors/CSharpTagHelperFieldDeclarationVisitor.cs

```diff
@@ -105,16 +106,16 @@ private void PreAllocateUnboundTagHelperAttributes(TagHelperChunk chunk)
             {
                 var attribute = chunk.Attributes[i];
                 var hasAssociatedDescriptors = chunk.Descriptors.Any(descriptor =>
-                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key)));
+                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Name)));
 
                 // If there's no descriptors associated or we're hitting a bound attribute a second time.
-                if (!hasAssociatedDescriptors || !boundAttributes.Add(attribute.Key))
+                if (!hasAssociatedDescriptors || !boundAttributes.Add(attribute.Name))
                 {
                     string preAllocatedAttributeVariableName = null;
 
-                    if (attribute.Value == null)
+                    if (attribute.Structure == HtmlAttributeStructure.Minimized)
                     {
-                        var preAllocatedAttributeKey = new TagHelperAttributeKey(attribute.Key, value: null);
```

Comment:
Might want a `Debug.Assert()` about here confirming `attribute.Value == null` and another confirming the opposite in the `else` block. For an `AttributeValueChunk`, `attribute.Structure == HtmlAttributeStructure.Minimized` and `attribute.Value == null` are redundant.